### DISCUSSION TITLE
UEFI binary

### DISCUSSION THREAD

**Discussion by smedir** [jglathe/linux_ms_dev_kit#20] [Q&A] [answered]
On amd64 efibootmgr is used to write boot entries to EFI bios. Is there something similar for arm64 in linux?

Reading the Linaro Debian install guide, your have to boot to shell.efi and then use the bcfg command to modify boot entries in nvram.

Where do I find the shell.efi app and associated documentation?

I am kind of tired of booting grub from a usb drive.

**Comment by jglathe** [comment]
Actually to get the HP X14 up first I installed systemd-boot. It worked - at a moment where grub didn't. Learned a few things:
- To successfully install systemd-boot you better have efivars support - it should be enabled in the v7 image for all currently supported x1e laptops.
- There is no need for shellaa64.efi, it should work with the installation of systemd-boot.
Of course I have also tried the [efi shell](https://github.com/pbatard/UEFI-Shell), seems that it can do better wirh some UEFI BIOSes than grub or sd-boot. I use it in combination with slbounce to boot the wdk to EL2. 
After having installed and booted Linux on the HP X14's internal SSD  though, I decided to upgrade to 24.10 (oracular). And its grub actually works with the built-in keyboard and can boot both Windows and Linux. So sd-boot was a very short intermezzo. Maybe you want to install grub on the local ssd? After creating an image of the original install, though.

**Comment by smedir** [comment]
Grub on the USB is actually working for me. It is also installed on the SSD
and should therefore work. I need to point the efi shell to that SSD grub,
I have to make an entry for it in the UEFI bios. currently there is only an
entry for the windows boot manager and an entry for USB/EFI partion.

**Comment by jglathe** [comment]
Oh, I guess I understand. There is no boot entry in the UEFI menu you coud choose from, right. My way to do this would be:
- boot from USB 
- unmount the EFI partition from USB
- mount the EFI partition from SSD instead
- run again `sudo grub-install`
- reboot and boot from SSD

This should at least add the boot entry in the UEFI BIOS if efivar is available. And usually, it should be the first/default boot entry too.

**Reply by smedir** [reply]
I am booted into Debian

EFI from NVME/SSD is correctly mounted.

`sudo grub-install` works correctly and all linux kernel and initrd.img entries are correctly created in the EFI partition.

The problem is no efivarfs.

/sys/firmware/efi/efivars is preset, but on ls, there are no entries in the directory.

efivarfs module is available.

````
modprobe efivarfs
root@SnapDragon: mount -t efivarfs efivarfs /sys/firmware/efi/efivars
mount: /sys/firmware/efi/efivars: fsopen system call failed: Operation not supported.
       dmesg(1) may have more information after failed mount system call.
````

I used your defconfig for building the kernel. And it has-

```
CONFIG_EFI_STUB=y
CONFIG_EFI=y
CONFIG_EFI_PARTITION=y
# EFI (Extensible Firmware Interface) Support
CONFIG_EFI_ESRT=y
CONFIG_EFI_PARAMS_FROM_FDT=y
CONFIG_EFI_RUNTIME_WRAPPERS=y
CONFIG_EFI_GENERIC_STUB=y
# CONFIG_EFI_ZBOOT is not set
CONFIG_EFI_ARMSTUB_DTB_LOADER=y
CONFIG_EFI_BOOTLOADER_CONTROL=y
CONFIG_EFI_CAPSULE_LOADER=y
# CONFIG_EFI_TEST is not set
# CONFIG_EFI_DISABLE_PCI_DMA is not set
CONFIG_EFI_EARLYCON=y
# CONFIG_EFI_DISABLE_RUNTIME is not set
# CONFIG_EFI_COCO_SECRET is not set
# end of EFI (Extensible Firmware Interface) Support
CONFIG_QCOM_QSEECOM_UEFISECAPP=y
CONFIG_FB_EFI=y
# CONFIG_RTC_DRV_EFI is not set
CONFIG_EFIVAR_FS=y
````

Your config did not have CONFIG_EFI_BOOTLOADER_CONTROL=y, however I did set it to yes. Despite all this, no EFI vars to point to grub entry from UEFI.

I am stumped for now. Any suggestions?

**Comment by jglathe** [comment, marked as answer]
For efivarfs to work (on grub install or sd-boot for example) the device must be recognized as compatible by the QSEECOM driver. I've done this [hack](https://github.com/jglathe/linux_ms_dev_kit/commit/a74c05edae358d44d427af63cf9bda469aa8a169) in newer versions, but it is not upstream yet.

**Reply by smedir** [reply]
Thanks. Have you commited the EC and the driver patch for the lid switch in 6.12. -rc?

If so, I can try and build that branch and see if things work with the lid switch and efibootmgr.

**Reply by jglathe** [reply]
I'm afraid not. I'm building on top of @jhovold's branches, and since I don't have a Yoga 7, can't verify.

**Reply by smedir** [reply]
And have you tried it :-). I don't really want to brick the laptop

**Reply by jglathe** [reply]
I don't have a Yoga 7, still approaching the odd audio topic cautiously. No, haven't tried on the X14 either.

**Reply by jglathe** [reply]
Oh misunderstood. Yes, I have tried these on all devices that I have. Snapdragon Dev Kit is pending, the others work.

**Reply by smedir** [reply]
Interestingly, your ubuntu installer, the current version you uploaded, does have efivars enabled, and efibootmgr works in showing current entries. I did not try to write with it yet :)

How is that working? How is the qseecom driver seeing compatibility there?

**Reply by jglathe** [reply]
The hack I referenced is in there. :man_shrugging:

**Reply by smedir** [reply]
So, just to clarify

The hack is in the Ubuntu branch
Is the hack commited in the 6.12 rc branch here? I guess I could go check but be easier to know from you directly

How about EC and the driver patch for the lid switch and power button- is that commited to 6.12.rc in your repo?

:smile:

**Reply by jglathe** [reply]
> The hack is in the Ubuntu branch Is the hack commited in the 6.12 rc branch here? I guess I could go check but be easier to know from you directly

yes 
 
> How about EC and the driver patch for the lid switch and power button- is that commited to 6.12.rc in your repo?

no, unless @jhovold carries it (but I guess he has no Yoga 7 either)

**Reply by smedir** [reply]
@jglathe 

THANK YOU!

ditched the USB. UEFI boot into grubaa64.efi, to debian. used your 6.12.rc3 kernel with the hack to write to UEFI with grub-install. Worked beautifully.

For others who may stumble across this!
